### DISCUSSION TITLE
Constarin Type Parameters As Numeric

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#1233] [General]
I was writing this struct:
```
    struct Range<T > where T : struct, IComparable
    {
        T Min { get; set; }
        T Max { get; set; }

        public Range(T min, T max)
        {
            Min = min;
            Max = max;
            _value = min;
        }

        T _value;
        T Value
        {
            get => _value;

            set
            {
                if (value.CompareTo(Max) == 1 || value.CompareTo(Min) == -1)
                    throw new ArgumentOutOfRangeException();

                _value = value;
            }
        }


        public static implicit operator T(Range<T> r)
        {
            return r.Value;
        }
    }

```
I want to use this range to deal with basic numeric data types and dates. Why there is no keywords for:
- numeric constraint.
- integer constraint (byte, sbyte, int, uint, short, ushort, long, ulong)
- floating constraint (float, double, decemal)
- value constraint (numerics and dates).
- Letiral constraint(Char and string).
- primitive constraint (numerics, dates, char, string)?

These basic types have common behavior, but they have't a base class or interface.

**Comment by HaloFour** [comment]
Why does this struct need any such constraints?  You're not performing any specialized arithmetic operations.

Also, DateTime wouldn't qualify as numeric, literal or primitive.

**Comment by ghost** [comment]
> Why does this struct need any such constraints? You're not performing any specialized arithmetic operations.

A range means a sequnce of continuous values. It implies arithmetic increment, which applies to numeric, date and char types.

> DateTime wouldn't qualify as numeric, literal or primitive.
It can make continuous range of dates. It is a numeric in this sense and date is reperesnted as double value.

**Comment by ghost** [comment]
besides, I needed this in the In-Range Data Types proposal #1198  .

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

> A range means a sequnce of continuous values. It implies arithmetic increment, which applies to numeric types and dates.

A range implies comparability between values, nothing more.  The potential continuous nature and lack of any universal unit makes arithmetic incrementing impossible.

> It can make continuous range of dates. It is a numeric in this sense and date is reperesnted as double value.

A `DateTime` is represented internally by a `ulong`, not a `double`.  And that `ulong` contains a bitmask flag indicating a type, which impacts that actual value of the `DateTime`.  What is the unit of `DateTime`?  Increment that 64-bit integer by 1 and you're moving the clock forward by 100 nanoseconds.  That's very likely not what you intend when you "add one" to a date.

**Comment by HaloFour** [comment]
I'll note that a `Range<T>` struct is already under design and consideration by the BCL team, to complement the proposed/championed range operator feature:

[System.Range.cs](https://github.com/dotnet/corefxlab/blob/d68a9a4cfae4c62a0cb05c204cf8d98d6876833a/src/System.Buffers.Experimental/System/Range.cs)

**Comment by ghost** [comment]
There is a deeference. The you can rename my range as ConstrainedNumeric. If the can add a Value property to the Range Class, then both ideas can be combined togather, but my ConstrainedNumeric is lightweight structure sutable for basic data types.

**Comment by orthoxerox** [comment]
@MohammadHamdyGhanem you'll need shapes for the constraints you want.

**Comment by ghost** [comment]
@HaloFour , @orthoxerox 
I posted a new proposal with a new approach:
https://github.com/dotnet/csharplang/issues/1387
